lib/auth/grpcserver.go

```diff
@@ -72,6 +90,124 @@ func (g *GRPCServer) SendKeepAlives(stream proto.AuthService_SendKeepAlivesServe
+				return
+			case statusUpdate := <-eventStream.Status():
+				if err := stream.Send(&statusUpdate); err != nil {
+					g.WithError(err).Debugf("Failed to send status update.")
+				}
+			}
+		}
+	}
+
+	closeStream := func(eventStream events.Stream) {
+		if err := eventStream.Close(auth.Context()); err != nil {
+			g.WithError(err).Warningf("Failed to flush close the stream.")
+		} else {
+			g.Debugf("Flushed and closed the stream.")
+		}
+	}
+
+	for {
+		request, err := stream.Recv()
+		if err == io.EOF {
+			return nil
+		}
+		if err != nil {
+			g.WithError(err).Debugf("Failed to receive stream request.")
+			return trail.ToGRPC(err)
+		}
+		if create := request.GetCreateStream(); create != nil {
+			if eventStream != nil {
+				return trail.ToGRPC(trace.BadParameter("stream is already created or resumed"))
+			}
+			eventStream, err = auth.CreateAuditStream(stream.Context(), session.ID(create.SessionID))
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			g.Debugf("Created stream: %v.", err)
+			go forwardEvents(eventStream)
+			defer closeStream(eventStream)
+		} else if resume := request.GetResumeStream(); resume != nil {
+			if eventStream != nil {
+				return trail.ToGRPC(trace.BadParameter("stream is already created or resumed"))
+			}
+			eventStream, err = auth.ResumeAuditStream(stream.Context(), session.ID(resume.SessionID), resume.UploadID)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			g.Debugf("Resumed stream: %v.", err)
+			go forwardEvents(eventStream)
+			defer closeStream(eventStream)
+		} else if complete := request.GetCompleteStream(); complete != nil {
+			if eventStream == nil {
+				return trail.ToGRPC(trace.BadParameter("stream is not initialized yet, can not complete"))
+			}
+			// do not use stream context to give the auth server finish the upload
+			// even if the stream's context is cancelled
+			err := eventStream.Complete(auth.Context())
+			g.Debugf("Completed stream: %v.", err)
+			if err != nil {
+				return trail.ToGRPC(err)
+			}
+			return nil
+		} else if flushAndClose := request.GetFlushAndCloseStream(); flushAndClose != nil {
+			if eventStream == nil {
+				return trail.ToGRPC(trace.BadParameter("stream is not initialized yet, can not flush and close"))
+			}
+			// flush and close is always done
+			return nil
+		} else if oneof := request.GetEvent(); oneof != nil {
+			if eventStream == nil {
+				return trail.ToGRPC(
+					trace.BadParameter("stream can not receive an event without first being created or resumed"))
+			}
+			event, err := events.FromOneOf(*oneof)
+			if err != nil {
+				g.WithError(err).Debugf("Failed to decode event.")
+				return trail.ToGRPC(err)
+			}
+			start := time.Now()
+			err = eventStream.EmitAuditEvent(stream.Context(), event)
+			if err != nil {
+				return trail.ToGRPC(err)
+			}
+			event.Size()
+			processed += int64(event.Size())
+			seconds := time.Since(streamStart) / time.Second
+			counter++
+			if counter%10000 == 0 {
```

Comment:
Might be more readable to declare `const logInterval = 10000` somewhere.

Reply:
fixed